packages/iris-grid/src/IrisGrid.tsx

```diff
@@ -2504,6 +2534,29 @@ class IrisGrid extends Component<IrisGridProps, IrisGridState> {
     onError(error);
   }
 
+  handleViewportUpdated(): void {
+    const { model } = this.props;
+    const { loadingText } = this.state;
+    const loadingMessage = 'Waiting for viewport...';
+
+    // pending and no timer already exists
+    if (model.isViewportPending && this.viewportLoadingTimeout === null) {
+      this.viewportLoadingTimeout = setTimeout(() => {
+        // still pending after timeout
+        if (model.isViewportPending) {
+          this.startLoading(loadingMessage, {
+            loadingCancelShown: false,
+            loadingBlocksGrid: false,
+          });
+        }
+        this.viewportLoadingTimeout = null;
+      }, VIEWPORT_LOADING_DELAY);
+    } else if (loadingText === loadingMessage && !model.isViewportPending) {
+      // extra conditions because timeout might get cleared by update
+      this.stopLoading();
+    }
```

Comment:
What you're doing here is throttling the function. This is a pretty common pattern. You can see `handleDownloadProgressUpdate` in this file (or `IrisGridTableModelTemplate.applyViewport`) for an example of a throttled function. But you could rewrite it like this:

```
componentWillUnmount() {
  ...
  // You at least need to stop the timer, you're not doing that currently.
  this.showViewportLoading.cancel();
  ...
}

showViewportLoading = throttle(
  (): void => {
    const { model } = this.props;
    const { loadingSpinnerShown } = this.state;
    if (model.isViewportPending && !loadingSpinnerShown) {
      // We only want to show the viewport loading if the viewport is still loading and we're not already showing a loader for something else
      this.startLoading("Waiting for viewport...", {
        loadingCancelShown: false,
        loadingBlocksGrid: false,
      });
    }
  },
  VIEWPORT_LOADING_DELAY
);


handleViewportUpdated(): void {
  this.showViewportLoading();
}

handleUpdate() {
  ...
  this.showViewportLoading.cancel();
  ...
}
```